common/transport/amqp/src/receiver_link.ts

```diff
@@ -0,0 +1,212 @@
+              this.emit('error', err);
+            }
+          },
+          attach: (callback) => {
+            this._fsm.transition('attaching', callback);
+          },
+          detach: (callback) => { this._safeCallback(callback); },
+          forceDetach: () => { return; },
+          accept: (message, callback) => callback(new errors.DeviceMessageLockLostError()),
+          reject: (message, callback) => callback(new errors.DeviceMessageLockLostError()),
+          abandon: (message, callback) => callback(new errors.DeviceMessageLockLostError())
+        },
+        attaching: {
+          _onEnter: (callback) => {
+            /*Codes_SRS_NODE_AMQP_RECEIVER_LINK_16_004: [The `attach` method shall use the stored instance of the `amqp10.AmqpClient` object to attach a new link object with the `linkAddress` and `linkOptions` provided when creating the `ReceiverLink` instance.]*/
+            debug('creating receiver with amqp10 for: ' + this._linkAddress);
+            this._amqp10Client.createReceiver(this._linkAddress, this._linkOptions)
+              .then((amqp10link) => {
+                debug('receiver object created by amqp10 for endpoint: ' + this._linkAddress);
+                if (this._fsm.state === 'attaching') {
+                  this._linkObject = amqp10link;
+                  /*Codes_SRS_NODE_AMQP_RECEIVER_LINK_16_006: [The `ReceiverLink` object should subscribe to the `detached` event of the newly created `amqp10` link object.]*/
+                  this._linkObject.on('detached', this._detachHandler);
+                  /*Codes_SRS_NODE_AMQP_RECEIVER_LINK_16_007: [The `ReceiverLink` object should subscribe to the `errorReceived` event of the newly created `amqp10` link object.]*/
+                  this._linkObject.on('errorReceived', this._errorHandler);
+                  this._linkObject.on('message', this._messageHandler);
+                  /*Codes_SRS_NODE_AMQP_RECEIVER_LINK_16_020: [The `attach` method shall call the `callback` if the link was successfully attached.]*/
+                  this._fsm.transition('attached', callback);
+                } else {
+                  debug('client forceDetached us already - cleaning up');
+                  amqp10link.forceDetach();
+                }
+                return null;
+              })
+              .catch((err) => {
+                debug('amqp10 failed to create receiver: ' + err.toString());
+                this._fsm.transition('detached', callback, err);
+              });
+          },
+          detach: (callback) => this._fsm.transition('detaching', callback),
+          forceDetach: (err) => this._fsm.transition('detached', undefined, err),
+          accept: (message, callback) => callback(new errors.DeviceMessageLockLostError()),
+          reject: (message, callback) => callback(new errors.DeviceMessageLockLostError()),
+          abandon: (message, callback) => callback(new errors.DeviceMessageLockLostError())
```

Comment:
defer until attached? (dangerous since we don't know if the attach is actually going to succeed).

Reply:
if the link has been detached, this can't succeed (the lock for the message was lost)

---
In reply to: [139543686](https://github.com/Azure/azure-iot-sdk-node/pull/83#discussion_r139543686) [](ancestors = 139543686)